multimodal/src/autogluon/multimodal/data/infer_types.py

```diff
@@ -159,103 +159,80 @@ def is_numerical_column(
+    elif isinstance(feature, list) and isinstance(feature[0], bytearray):
+        image_feature = feature[0]
+    else:
+        return image_type
 
-    Returns
-    -------
-    Whether the column is an image-path column.
-    """
-    sample_num = min(len(data), sample_n)
-    data = data.sample(n=sample_num, random_state=0)
-    data = data.apply(lambda ele: str(ele).split(";")).tolist()
-    failure_count = 0
-    for image_paths in data:
-        success = False
-        for img_path in image_paths:
-            try:
-                with PIL.Image.open(img_path) as img:
-                    pass
-                success = True
-                break
-            except:
-                pass
-        if not success:
-            failure_count += 1
-    failure_ratio = failure_count / sample_num
+    try:
+        with PIL.Image.open(image_feature) as img:
+            image_type = IMAGE_PATH
+    except:
+        try:
+            with PIL.Image.open(BytesIO(image_feature)) as img:
+                image_type = IMAGE_BYTEARRAY
+        except:
+            return image_type
 
-    # Tolerate high failure rate in case that many image files may be corrupted.
-    if failure_ratio <= 0.9:
-        if failure_ratio > 0:
-            warnings.warn(
-                f"Among {sample_num} sampled images in column '{col_name}', "
-                f"{failure_ratio:.0%} images can't be open. "
-                "You may need to thoroughly check your data to see the percentage of missing images, "
-                "and estimate the potential influence. By default, we skip the samples with missing images. "
-                "You can also set hyperparameter 'data.image.missing_value_strategy' to be 'zero', "
-                "which uses a zero image to replace any missing image.",
-                UserWarning,
-            )
-        return True
-    else:
-        return False
+    return image_type
 
 
-def is_imagebytearray_column(
+def is_image_column(
     data: pd.Series,
     col_name: str,
     sample_n: Optional[int] = 500,
 ) -> bool:
     """
-    Identify if a column contains valid image bytearray.
+    Identify if a column is one image column.
+    Here it counts the failures when trying PIL.Image.open() on a sampled subset.
+    If over 90% attempts fail, this column isn't an image column.
 
     Parameters
     ----------
     data
         One column of a multimodal pd.DataFrame for training.
     col_name
         Name of column.
-    sample_n
-        Number of sample images to open for sanity check.
 
     Returns
     -------
-    Whether the column is an image-bytearray column.
+    Whether the column is an image column.
     """
     sample_num = min(len(data), sample_n)
     data = data.sample(n=sample_num, random_state=0)
-    data = data.tolist()
+    image_type = get_image_feature_type(data.iloc[0])
+    if image_type == IMAGE_PATH:
+        data = data.apply(lambda ele: str(ele).split(";")).tolist()
```

Comment:
Please use vectorized form: `data.str.split(';').tolist()`

Reply:
> 



> > > We need to distinguish `image_path` and `image_bytearray` in the column types. These image sub-types would be used in df preprocessor and image processor. The current modification breaks the original logic.
> > 
> > 
> > The problem is that there is a difference of usage between training and inference time. We can train using images on disk and then use bytearrays during the inference. Saving files to a file system during the inference time creates additional security overhead.
> 
> I agree that we need to support using image_path in training and image_bytearray in inference. To do so, we can infer the image column types during inference and see whether the subtype changes. If we detect subtype changes, we can modify the `_column_types` in df preprocessor to reflect them. There is no need to change the internal logic among infer_types, df preprocessor, and processor.

Thanks for the suggestion. I'm working on a quick POC for this idea, and hopefully we can still catch the release.